gate-oauth2/src/main/java/com/netflix/spinnaker/gate/security/oauth2/SpinnakerUserInfoTokenServices.java

```diff
@@ -0,0 +1,304 @@
+import org.springframework.security.oauth2.common.OAuth2AccessToken;
+import org.springframework.security.oauth2.common.exceptions.InvalidTokenException;
+import org.springframework.security.oauth2.provider.OAuth2Authentication;
+import org.springframework.security.oauth2.provider.OAuth2Request;
+import org.springframework.security.oauth2.provider.token.ResourceServerTokenServices;
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken;
+
+/**
+ * ResourceServerTokenServices is an interface used to manage access tokens. The
+ * UserInfoTokenService object is an implementation of that interface that uses an access token to
+ * get the logged in user's data (such as email or profile). We want to customize the Authentication
+ * object that is returned to include our custom (Kork) User.
+ */
+@Slf4j
+public class SpinnakerUserInfoTokenServices implements ResourceServerTokenServices {
+
+  @Autowired private ResourceServerProperties sso;
+  @Autowired private UserInfoTokenServices userInfoTokenServices;
+  @Autowired private CredentialsService credentialsService;
+  @Autowired protected OAuth2SsoConfig.UserInfoMapping userInfoMapping;
+  @Autowired protected OAuth2SsoConfig.UserInfoRequirements userInfoRequirements;
+  @Autowired private PermissionService permissionService;
+  @Autowired private Front50Service front50Service;
+
+  @Autowired(required = false)
+  private SpinnakerProviderTokenServices providerTokenServices;
+
+  @Autowired private AllowedAccountsSupport allowedAccountsSupport;
+  @Autowired private FiatClientConfigurationProperties fiatClientConfigurationProperties;
+  @Autowired private Registry registry;
+
+  @Autowired(required = false)
+  @Qualifier("spinnaker-oauth2-group-extractor")
+  private BiFunction<String, Map, List<String>> groupExtractor;
+
+  private RetrySupport retrySupport = new RetrySupport();
+
+  @Override
+  public OAuth2Authentication loadAuthentication(final String accessToken)
+      throws AuthenticationException, InvalidTokenException {
+    OAuth2Authentication oAuth2Authentication =
+        userInfoTokenServices.loadAuthentication(accessToken);
+
+    final Map<String, Object> details =
+        (Map<String, Object>) oAuth2Authentication.getUserAuthentication().getDetails();
+
+    log.debug("UserInfo details: " + details);
+
+    boolean isServiceAccount = isServiceAccount(details);
+    if (!isServiceAccount) {
+      if (!hasAllUserInfoRequirements(details)) {
+        throw new BadCredentialsException("User's info does not have all required fields.");
+      }
+
+      if (providerTokenServices != null
+          && !providerTokenServices.hasAllProviderRequirements(accessToken, details)) {
+        throw new BadCredentialsException(
+            "User's provider info does not have all required fields.");
+      }
+    }
+
+    final String username = safeGet(details.get(userInfoMapping.getUsername()));
```

Comment:
Is this really the equivalent of
```
    def username = details[userInfoMapping.username] as String
```
As in, does it handle a null value for `userInfoMapping.getUsername()` the same way?  And, does username end up as the empty string in the same cases that `safeGet` returns.

Reply:
the equivalent code is

```
final String username = Optional.ofNullable((String) details.get(userInfoMapping.getUsername()))
                          .orElse("null");
```

Reply:
Implemented toStringOrNull, which is a groovy equivalent functionality for the above case.

Reply:
https://github.com/spinnaker/gate/pull/1871/commits/51348ba555a2cc53a433626517d4c706d0fa4e49